lib/exercises.rb

```diff
@@ -2,21 +2,47 @@
 
 # This method will return an array of arrays.
 # Each subarray will have strings which are anagrams of each other
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: o(n)
+# Space Complexity: o(n)
 
 def grouped_anagrams(strings)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
+  # raise NotImplementedError, "Method hasn't been implemented yet!"
+  hash = {}
+  strings.each do |string|
+    sorted_string = string.split("").sort.join
+    if hash[sorted_string]
+      hash[sorted_string] << string
+    else 
+      hash[sorted_string] = [string]
+    end
+  end
+  return hash.values
 end
 
+
 # This method will return the k most common elements
 # in the case of a tie it will select the first occuring element.
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: o(n)
+# Space Complexity: o(n)
 def top_k_frequent_elements(list, k)
-  raise NotImplementedError, "Method hasn't been implemented yet!"
-end
+  # raise NotImplementedError, "Method hasn't been implemented yet!"
+  hash = {}
+  output = []
+  return list if list.empty?
 
+  list.each do |element|
+    if hash[element]
+      hash[element] += 1
+    else
+      hash[element] = 1
+    end
+  end
+  
+  k.times do |element|
+    output << hash.sort[element][0]
+  end
```

Comment:
I believe `hash.sort` sorts elements by the key instead of the value, so this will return the elements sorted by the keys instead of the frequency of their occurrence in the original array.

It's happening to pass the tests, but it's not working in general.

Consider this:

```ruby

    it "will work for [3, 3, 3, 3, 3, 1, 1, 2, 2, 2]" do
      list = [3, 3, 3, 3, 3, 1, 1, 2, 2]
      k = 2

      answer = top_k_frequent_elements(list, k)

      expect(answer.sort).must_equal [2, 3]
    end
```

Also even if this did work, you're sorting the hash each iteration of the loop... which is excessive.